Removed TARGET_CY8C62XX; Added ARMC6 support; Updated PSA targets

hennadiykytsun opened the pull request:
### Description

Release: feature

Added ARMC6 support (fixed issue #9830 )
Updated target implementation (already supported target) - removed CSP target (TARGET_62XX) and used existing BSP targets instead.
Updated version of startup (+linker scripts), flash and ipc drivers to support PSA targets with the shared memory space.
Removed custom IPC configuration - used default IPC configuration from startup instead.


### Pull request type

    [ ] Fix
    [x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@cydriftcloud @cy-opm @vmedcy

0xc0170 reviewed (changes requested):
Request above ^^

vmedcy commented on targets/targets.json:
```diff
@@ -8172,11 +8172,15 @@
     "CY8CKIT_062_WIFI_BT_PSA": {
-        "inherits": ["NSPE_Target", "CY8CKIT_062_WIFI_BT"],
-        "extra_labels_add": ["PSA", "MBED_SPM"],
+        "inherits": ["NSPE_Target", "MCU_PSOC6_M4"],
```
`CY8CKIT_062_WIFI_BT_PSA` no longer inherits from `CY8CKIT_062_WIFI_BT`. I do not think this is correct, PSA target should inherit from non-PSA target instead of duplicating all attributes.

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/7239375953203437a4c880ac65623ddbebf0a093

vmedcy replied:
CY8CKIT_062_WIFI_BT_PSA now inherits from CY8CKIT_062_WIFI_BT, as intended.

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_4343W/device/system_psoc6.h:
```diff
@@ -539,7 +544,11 @@ extern "C" {
 */
-extern void SystemInit(void);
+#if defined(__ARMCC_VERSION)
+    extern void SystemInit(void) __attribute__((constructor));
```
remove __attribute__((constructor))

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/d4f433a8cdab56b5ef8994e0c943de0032ae85f8

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_BLE/device/system_psoc6.h:
```diff
@@ -539,7 +544,11 @@ extern "C" {
 */
-extern void SystemInit(void);
+#if defined(__ARMCC_VERSION)
+    extern void SystemInit(void) __attribute__((constructor));
```
remove attribute

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/d4f433a8cdab56b5ef8994e0c943de0032ae85f8

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_WIFI_BT/device/system_psoc6.h:
```diff
@@ -539,7 +544,11 @@ extern "C" {
 */
-extern void SystemInit(void);
+#if defined(__ARMCC_VERSION)
+    extern void SystemInit(void) __attribute__((constructor));
```
remove attribute

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/d4f433a8cdab56b5ef8994e0c943de0032ae85f8

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_WIFI_BT_M0_PSA/device/system_psoc6.h:
```diff
@@ -0,0 +1,674 @@
+* \{
+*/
+#if defined(__ARMCC_VERSION)
+    extern void SystemInit(void) __attribute__((constructor));
```
remove attribute

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/d4f433a8cdab56b5ef8994e0c943de0032ae85f8

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_WIFI_BT_PSA/device/system_psoc6.h:
```diff
@@ -0,0 +1,674 @@
+* \{
+*/
+#if defined(__ARMCC_VERSION)
+    extern void SystemInit(void) __attribute__((constructor));
```
remove attribute

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/d4f433a8cdab56b5ef8994e0c943de0032ae85f8

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CMOD_062_4343W/device/system_psoc6.h:
```diff
@@ -539,7 +544,11 @@ extern "C" {
 */
-extern void SystemInit(void);
+#if defined(__ARMCC_VERSION)
+    extern void SystemInit(void) __attribute__((constructor));
```
remove attribute

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/d4f433a8cdab56b5ef8994e0c943de0032ae85f8

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CYW943012P6EVB_01/device/system_psoc6.h:
```diff
@@ -0,0 +1,674 @@
+* \{
+*/
+#if defined(__ARMCC_VERSION)
+    extern void SystemInit(void) __attribute__((constructor));
```
remove attribute

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/d4f433a8cdab56b5ef8994e0c943de0032ae85f8

orenc17 reviewed (changes requested):
Remove `__attribute__((constructor))` from SystemInit()
this has caused problems in the past

orenc17 commented on targets/targets.json:
```diff
@@ -8172,11 +8172,15 @@
     "CY8CKIT_062_WIFI_BT_PSA": {
-        "inherits": ["NSPE_Target", "CY8CKIT_062_WIFI_BT"],
-        "extra_labels_add": ["PSA", "MBED_SPM"],
+        "inherits": ["NSPE_Target", "MCU_PSOC6_M4"],
```
please restore the previous statae
PSA targets should always inherit from the closest match to the real target, for maintenance reasons

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/7239375953203437a4c880ac65623ddbebf0a093

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_WIFI_BT_M0_PSA/spm_api.h:
```diff
@@ -0,0 +1,30 @@
+/* Copyright (c) 2017-2018 ARM Limited
```
conflicts with `hal\spm_api.h` please remove

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_WIFI_BT_PSA/spm_api.h:
```diff
@@ -0,0 +1,30 @@
+/* Copyright (c) 2017-2018 ARM Limited
```
conflicts with `hal\spm_api.h` please remove

hennadiykytsun replied:
Fixed: removed this header

orenc17 reviewed (changes requested):
I really don't understand why we need so much file duplication
the `TARGET_CY8CKIT_062_WIFI_BT_M0_PSA` & `TARGET_CY8CKIT_062_WIFI_BT_PSA` should contains only PSA related files

there is no need to have duplicates

@alzix can you take a look as well?

hennadiykytsun committed:
PSoC6: Remove TARGET_CY8C62XX CSP directory

GeneratedSource folders are BSP specific. No parts of the kit BSP can be reused
as generic chip support package. Remove TARGET_CY8C62XX directory,
and use flat BSP inheritance model:

MCU_PSOC6 -> MCU_PSOC6_M4 -> CY8CKIT_062_WIFI_BT
MCU_PSOC6 -> MCU_PSOC6_M0 -> CY8CKIT_062_WIFI_BT_M0

hennadiykytsun committed:
PSOC6: remove custom IPC configuration for PSA

* Update PDL startup driver to version 2.40
* Update linker scripts and startup assembly
* Remove custom IPC configuration from PSA initialization:
  use default IPC configuration provided by low-level startup code.

hennadiykytsun committed:
PSOC6: update version of PDL flash driver

Flash driver 3.30:
Moved ipcWaitMessageStc structure to the RAM section called ".cy_sharedmem"
Added support Secure Boot devices
Moved CY_FLASH_EFFECTIVE_PAGE_SIZE to flash_api.c (the macro is Mbed specific).

hennadiykytsun committed:
PSOC6: update version of PDL IPC driver

hennadiykytsun committed:
PSOC6: add ARMC6 support (fix issue #9830)

Update PDL syslib driver to 2.30.
Update startup assembly and linker scripts.

hennadiykytsun committed:
FUTURE_SEQUANA: reuse supported_toolchains from TARGET_PSOC6

hennadiykytsun committed:
PSOC6: rebuild CY8CKIT_062_WIFI_BT_PSA prebuilt HEX files

hennadiykytsun committed:
PSOC6: remove __attribute__((constructor)) from SystemInit

vmedcy commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_WIFI_BT_PSA/prebuilt/README.md:
````diff
@@ -7,7 +7,7 @@ These images were compiled by the following command:
 ```
 mbed compile -m CY8CKIT_062_WIFI_BT_M0_PSA -t GCC_ARM --profile release -N psa_release_1.0
-mbed test --compile -m CY8CKIT_062_WIFI_BT_M0_PSA -t GCC_ARM --profile debug -n *psa-spm*,*psa-crypto_access_control
+mbed test --compile -m CY8CKIT_062_WIFI_BT_M0_PSA -t GCC_ARM --profile release -n *psa-spm*,*psa-crypto_access_control
````
Since https://github.com/ARMmbed/mbed-os/commit/04c58031310c6eb214600eb1cac4368e3454c091, debug binaries no longer fit into CM0+ address space, hence switched to release profile.

orenc17 replied:
please use `python tools/psa/release.py -m CY8CKIT_062_WIFI_BT_M0_PSA ` instead of the 2 coomands

orenc17 replied:
please replace both commands with `python tools/psa/release.py -m CY8CKIT_062_WIFI_BT_M0_PSA `

hennadiykytsun replied:
Fixed in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/83275da0cf4d25a4b8da3d9ed7f195188e76de69 commit

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_BLE/device/system_psoc6_cm4.c:
```diff
@@ -43,6 +43,10 @@
     #endif /* defined(CY_DEVICE_PSOC6ABLE2) */
 #endif /* !defined(CY_IPC_DEFAULT_CFG_DISABLE) */
 
+#if defined(COMPONENT_SPM_MAILBOX)
```
please remove this code until you actually add the BLE kit as a PSA target

vmedcy replied:
See https://github.com/ARMmbed/mbed-os/pull/10308#discussion_r275443336

hennadiykytsun replied:
Updated readme.me in the https://github.com/ARMmbed/mbed-os/pull/10308/commits/83275da0cf4d25a4b8da3d9ed7f195188e76de69 commit

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_BLE/device/system_psoc6_cm4.c:
```diff
@@ -277,13 +281,22 @@ void SystemInit(void)
 *******************************************************************************/
 void mbed_sdk_init(void)
 {
+#if !defined(COMPONENT_SPM_MAILBOX)
```
please remove this code until you actually add the BLE kit as a PSA target

vmedcy replied:
See https://github.com/ARMmbed/mbed-os/pull/10308#discussion_r275443336

orenc17 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_BLE/device/system_psoc6_cm4.c:
```diff
@@ -277,13 +281,22 @@ void SystemInit(void)
     /* Placed here as it must be done after proper LIBC initialization. */
     SystemInit();
 
+#if defined(COMPONENT_SPM_MAILBOX)
```
please remove this code until you actually add the BLE kit as a PSA target

vmedcy replied:
No, system_psoc6_cm4.c is deployed from internal PDL development repo, there is a desire to keep the file identical across supported PSoC6 BSPs. It is put into the BSP target directories instead of psoc6pdl directory to better align with CMSIS template (the file can changed when creating custom BSPs).

orenc17 reviewed (changes requested):
Looks fine..
i will run the PSA tests on my board tomorrow to verify that nothing is broken

in the meantime, please see my remarks

hennadiykytsun committed:
CY8CKIT_062_WIFI_BT_PSA: mention tools/psa/release.py in README

orenc17 reviewed (approved):
Tested with #10413
